### PR TITLE
feat: record detailed trace of IO operations

### LINKED ISSUE #7053: Collect detailed informatoin about IO related work in an intermediate format
Use an intermediate representation of database accesses and what caused them. This trace of IO operations should be easy to generate from all `neard` commands and also from the estimator.

Motivation:

*   Quickly understand the DB pressure any workload of interest generates.
*   Store intermediate traces to perform indirect estimations of IO cost.

### PR DESCRIPTION
**Pull request by jakmeier**
Allows both neard and the runimte-parameter-estimator to collect a
human-readable trace of executed storage and DB operations.

The tracing infrastructure is used here to collect spans and events
at runtime. The new subscriber layer `IoTraceLayer` can be enabled
for neard and for the runtime estimator, with the flag
`--record-io-trace=my_file.log`.

A new tracing target "io_tracer" has been added for events that are
only of interest to this new subscriber layer. And all such event are
emitted on the Trace level.

To avoid affecting release mode neard, I added the feature
`tracing/release_max_level_debug` to store and vm-logic.
This means the new feature only works on non-release builds.

Some existing spans are also reused, those are in the targets
"host-function", "store", "vm_logic", "runtime".

The recorded trace can be used to analyse DB access patterns.

In a next step, the estimator could also use such a trace to estimate
IO costs indirectly.

Resolves #7053, which also contains more details on the output format.

### CODE REVIEW

**Comment by jakmeier** [comment]
For the why and what it looks like in action, take a look at https://github.com/near/nearcore/issues/7053#issuecomment-1162393161 and https://github.com/near/nearcore/issues/7060#issuecomment-1162412350.

**Review by nagisa — COMMENTED** [review]
Having looked at this I’m not quite sure if tracing is a right tool for this task. Or at least isn’t being utilized the way it is meant to be.

On one hand it is nice that it is global, on another this is trying to make it do things `tracing` was not really designed for – such as computing what seems to be a metric-esque thing out of the events.

My, admittedly somewhat of a kneejerk, reaction is one or some of the:

1. Could this feature consume events from the other end of the already-existing opentelemetry service?
2. Maybe this could be made much simpler by moving some logic around metric calculation onto the code that will consume the events?
3. And in doing so enable us to use a `tracing_subscriber::fmt` layer with a JSON format to avoid rolling our own serialization format over rolling our own? Those can always be transformed into a human readable form, and doing so seems much easier than going the other way.

That kind of thoughts, anyway. I did try to review this implementation assuming we want to go in the proposed direction no matter what. There are a lot places where the implementation seems limited by the lack of the integration with `valuable`, for example, which makes this code look more off than it strictly needs to be…

**Comment by jakmeier** [comment]
Thank you @nagisa for your thoughts! I was hoping for this kind of feedback. The current version has just grown as it was the easiest for me to explore what data we actually need and what format is useful for replaying and also manual inspection.

>     1. Could this feature consume events from the other end of the already-existing opentelemetry service?

Maybe. I considered it shortly but abandoned the idea because it seemed that requires some listening application somewhere, even for estimator runs. Also, I am worried this will be a bit too much traffic hundreds of thousands of events per second. But to be honest, I am just not familiar with opentelemetry and cannot judge if it is the right tool.


>     2. Maybe this could be made much simpler by moving some logic around metric calculation onto the code that will consume the events?

Hm, keeping the production code as clean as possible is a high priority here. Spilling it with metric accumulation code would have to be done very carefully. But I can give it a try.

>     3. And in doing so enable us to use a `tracing_subscriber::fmt` layer with a JSON format to avoid rolling our own serialization format? JSON events can always be transformed into a human readable form, and doing so seems much easier than going the other way.

One of my earlier iterations of this used the JSON output. Maybe I was doing something wrong but I found the span representation in JSON format utterly useless for my use case. :sweat_smile: (Looking at the trace as human, or replaying operations in a streaming matter, knowing the result of operations without look-ahead.) So I decided to rather skip yet another intermediate layer and produce something that is immediately useful to humans and machines. 


### Alternative idea
Perhaps a cleaner idea would be to run our own global metric collection service. It could also store a stack of spans, a map of key-value pairs, and a map of counter values. And then probably add a feature that makes everything a NOOP if not enabled.

@nagisa I am interested to hear your gut feeling here. Which of all the options (Opentelemetry / accumulate in code + JSON / alternative home-grown metrics service) would you think is most promising to explore?

File: core/store/Cargo.toml [hunk 1]
```diff
@@ -22,7 +22,7 @@ num_cpus = "1.11"
 rand = "0.7"
 strum = { version = "0.24", features = ["derive"] }
 fs2 = "0.4"
-tracing = "0.1.13"
+tracing = { version = "0.1.13", features = ["release_max_level_debug"] }
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/store/Cargo.toml:25]: I don't think it skews the results in any way, since timing is NOT something I want to record in this way. This trace is to have a record of _what_ happens exactly during execution. To get accurate timing info of the full system, the suggested feature should be turned off anyway. And to get timing info in controlled isolation, the traces would have to be replayed separately on the estimator.

That said, I think a new feature and some `cfg!` might be a better solution after all. I was just trying to avoid adding more crate features.

**Comment by nagisa** [comment]
> Perhaps a cleaner idea would be to run our own global metric collection service. It could also store a stack of spans, a map of key-value pairs, and a map of counter values. And then probably add a feature that makes everything a NOOP if not enabled.

There _will_ be a distributed event and span collector in a form of Grafana Tempo via otel, I believe. cc @nikurt 

---

> One of my earlier iterations of this used the JSON output. Maybe I was doing something wrong but I found the span representation in JSON format utterly useless for my use case. sweat_smile (Looking at the trace as human, or replaying operations in a streaming matter, knowing the result of operations without look-ahead.)

The problem of looking at machine-readable output as a human is better solved by having some tooling in place to make it more readable and browsable (possibly as part of the upcoming tool that summarizes the results). This is, sort of, the same idea as storing the logs plaintext vs doing so with journald. Yes, you might be able to construct a `grep` invocation to filter out logs from a specific service, but journald gives you a nice dedicated flag for it.

I can see why the way JSON presents span might not be super useful though. In opentelemetry those are referenced by unique IDs so it is easy to cross-reference and re-construct the stack of spans and events’ relationship to them at a future time. JSON output seems like it doesn’t enable that sort of thing as easily…

---

> > 2. Maybe this could be made much simpler by moving some logic around metric calculation onto the code that will consume the events?
>
> Hm, keeping the production code as clean as possible is a high priority here. Spilling it with metric accumulation code would have to be done very carefully. But I can give it a try.

There’s probably a misinterpretation somewhere, probably because I described two different alternate approaches. What I meant here specifically was: instead of dealing with `count` in the `IoTraceLayer`, emit the `count` events into your file verbatim and handle the counting in whatever tooling consumes those events. This wouldn’t make for immediately useful textual output, but it would definitely result in a simpler neard.

That said, if the _only_ reason why these `count` events are instrumented is to derive metrics,  then as I said before, a tool oriented towards _metrics_ is probably more appropriate.

The alternative with `record` that I mentioned elsewhere outside of the quoted section would end up approximately the same complexity wise as `event!` is right now. Something like this, probably:

```rust
tracing::Span::current().record("iotrace.count", "some_event");
```

---

> Maybe. I considered it shortly but abandoned the idea because it seemed that requires some listening application somewhere, even for estimator runs. Also, I am worried this will be a bit too much traffic hundreds of thousands of events per second.

Yes, opentelemetry does require a collector process to receive the otlp data on the other end of the socket. Maybe it is premature of me to predict this at the moment, but I do anticipate that it will become an important part of of our toolbox for local development as well not too shortly.

As for 100kHz+ events – is it really the amount of data necessary for the functionality? This sounds like it’d be quite a lot to deal with even for the most micro-optimized subscribers. Being able to handle an event in ~100ns is pretty hard – at those time scales decrementing an `Arc` reference count can become a significant cost centre, let alone serialization and other such tasks. Multiplying this by 100k will leave you at significant percentages of overhead just for handling instrumentation…

In practice I haven’t seen otel come up as a significant problem, at least not in the use-cases where most alternative subscribers weren’t already a problem too. That said it is definitely worth to try benchmarking it ^^

**Comment by jakmeier** [comment]
@nagisa Thank you for the detailed thoughts! I will take your inputs and see how I can make this work better. :)

Regarding the 100kHz+ events, I came up with that number by looking at what I emit today. That is, counting things like trie node cache hits, which are supposed to finish in 2us according to the gas cost. That would be 500kHz. But in the best case, using a better fitting tool to count metrics, the overhead can be reduced to a single integer inc. (Probably this doesn't even need to be an atomic int.)

To be clear, although I don't think it will be necessary, I would actually be fine with 10x slower execution when this type of instrumentation is enabled. My goal is to extract very fine-grained information about what the system does to handle requests. This is not something we should ever run in production.

File: core/o11y/src/io_tracer.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            "col" => self.col = Some(value.to_owned()),
+            "storage_op" => {
+                let op = match value {
+                    "write" => StorageOp::Write,
+                    "read" => StorageOp::Read,
+                    _ => StorageOp::Other,
+                };
+                self.t = Some(IoEventType::StorageOp(op));
+            }
+            // GET operation has a `Debug` printed `Option<usize>` for size.
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs:209]: What you are writing here makes sense. But can I set `[field::Empty](https://docs.rs/tracing/latest/tracing/field/struct.Empty.html)` dynamically based on runtime values? 
```
```

File: core/o11y/src/io_tracer.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        span.extensions_mut().replace(OutputBuffer(vec![]));
+        span.extensions_mut().replace(SpanInfo::default());
+    }
+
+    fn on_event(&self, event: &tracing::Event<'_>, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        if event.metadata().target() == "io_tracer" {
+            // Events specifically added to add more info to spans in IO Tracer.
+            // Marked with `target: "io_tracer"`.
+            if let Some(span) = ctx.event_span(event) {
+                event.record(span.extensions_mut().get_mut::<SpanInfo>().unwrap());
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs:68]: Yeah I'm now walking up parents spans. I wanted to implement it with delayed field records but the requirement to define all fields statically on all possible parent spans lead me to roll back and not do that.

File: core/o11y/src/io_tracer.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+struct SpanInfo {
+    key_values: Vec<String>,
+    counts: HashMap<String, u64>,
+}
+
+impl<S: Subscriber + for<'span> LookupSpan<'span>> Layer<S> for IoTraceLayer {
+    fn on_enter(&self, id: &span::Id, ctx: tracing_subscriber::layer::Context<'_, S>) {
+        let span = ctx.span(id).unwrap();
+        span.extensions_mut().replace(OutputBuffer(vec![]));
+        span.extensions_mut().replace(SpanInfo::default());
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs:60]: 1. Ok, I changed that. But for my education, is there a reason why `on_new_span` is preferred over `on_enter`? My thinking was if spans are reused somewhere, clearing every time it is entered should be the more robust solution. But then again, I have very little experience with `tracing` and how it is typically used.
2. I can't seem find `on_register` or something with a similar name anywhere, where is it defined? [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) doesn't have it, nor can I find it anywhere else in the docs for the involved tracing crates.

File: runtime/runtime/src/lib.rs [hunk 5]
```diff
@@ -1254,6 +1255,7 @@ impl Runtime {
                 receipt_id = %receipt.receipt_id,
                 node_counter = ?state_update.trie.get_trie_nodes_count())
             .entered();
+            tracing::trace!(target: "io_tracer", predecessor = %receipt.predecessor_id, receiver = %receipt.receiver_id, id = %receipt.receipt_id);
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/runtime/src/lib.rs]: This makes sense and I updated it accordingly. Now I generally include span fields as well. (With the exception of those that I noticed being formatted strings rather than simple key-value pairs, which were too verbose for human readability and would be hard to parse for computers.)

File: core/o11y/src/io_tracer.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nagisa) [reply core/o11y/src/io_tracer.rs]: > But for my education, is there a reason why on_new_span is preferred over on_enter? My thinking was if spans are reused somewhere, clearing every time it is entered should be the more robust solution. But then again, I have very little experience with tracing and how it is typically used.

For instrumented `Future`s the span will be `enter`ed every single time a future is polled. This is can be an extremely frequent operation, so ideally it should be relatively extremely cheap so that the overhead on async code isn’t outrageous. Outside of that, not dealing with the overhead for spans that are manually entered and exited multiple times is also nice. Sorry for not explaining it in the first comment.

> I can't seem find on_register or something with a similar name anywhere, where is it defined? [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) doesn't have it, nor can I find it anywhere else in the docs for the involved tracing crates.

Sorry, my bad. I meant `on_record` here. In practice what I’m saying here is that

```
let span = tracing::info_span!("name", size = tracing::field::Empty).entered();
span.record("size", &124);
```

should produce largely the same results as

```
let span = tracing::info_span!("name", size = &124).entered();
```

File: core/o11y/src/io_tracer.rs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            writeln!(out, "{span_line}").unwrap();
+            for BufferedLine { indent, output_line } in exiting_buffer.drain(..) {
+                writeln!(out, "{:indent$}{output_line}", "").unwrap();
+            }
+        }
+    }
+}
+
+impl IoTraceLayer {
+    pub fn new(file: Mutex<File>) -> Self {
```
[the remark this answers is not in the document]
Comment (nagisa) [reply core/o11y/src/io_tracer.rs:131]: This is independent of `fmt::Json`. `MakeWriter` will construct you instances of the `io::Write` trait, making the overall subscriber generic over how it writes things to… wherever, and composable with various pre-existing high quality writer implementations such as [`NonBlocking`](https://docs.rs/tracing-appender/latest/tracing_appender/non_blocking/struct.NonBlocking.html).

Alternatively you could potentially make it generic over `io::Write` for the time being, if only to simplify the implementation effort. But you would need the `MakeWriter` if you wanted to have an ability to `lock` the mutex once per event/span rather than for each `write` operation.

[the remark this answers is not in the document]
Comment (nagisa) [reply core/o11y/src/io_tracer.rs:131]: See in particular [`impl MakeWriter for Mutex<W>`](https://docs.rs/tracing-subscriber/0.3.9/src/tracing_subscriber/fmt/writer.rs.html#792-801)

File: runtime/near-vm-logic/src/logic.rs [hunk 7]
```diff
@@ -2424,8 +2436,20 @@ impl<'a> VMLogic<'a> {
         let nodes_before = self.ext.get_trie_nodes_count();
         let read = self.ext.storage_get(&key);
         let nodes_delta = self.ext.get_trie_nodes_count() - nodes_before;
+        #[cfg(feature = "io_trace")]
+        let (tn_mem_reads, tn_db_reads) = (nodes_delta.mem_reads, nodes_delta.db_reads);
```
Comment (nagisa) [inline runtime/near-vm-logic/src/logic.rs:2440]: Style nit: Inline these variables to avoid having two distinct `cfg`s?

**Comment by nagisa** [comment]
Okay, got it. Thanks for exploring the other options. Since the `tracing` implementation isn’t particularly onerous, it does solve a problem and the other options aren’t obviously better nor implementable in a straightforward way, I think it is fine for us to just keep tracing-based implementation then.

File: core/o11y/src/io_tracer.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs]: Thank you for the explanations! This helps a lot.

I added the implementation for records as well.

File: core/o11y/src/io_tracer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs]: Oh, I am just stupid then. :sweat_smile:  Apparently I didn't test the obvious solution of just using a plain `Option` properly. (I did try it but somehow something didn't work out then.)

Anyway, it works now.

File: core/store/src/trie/trie_storage.rs [hunk 8]
```diff
@@ -241,8 +241,14 @@ impl TrieStorage for TrieCachingStorage {
         // Try to get value from shard cache containing most recently touched nodes.
         let mut guard = self.shard_cache.0.lock().expect(POISONED_LOCK_ERR);
         let val = match guard.get(hash) {
-            Some(val) => val.clone(),
+            Some(val) => {
+                #[cfg(feature = "io_trace")]
+                tracing::trace!(target: "io_tracer", count = "shard_cache_hit");
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/store/src/trie/trie_storage.rs]: I really like this. I added `near_o11y::io_trace!` for this purpose.

File: runtime/near-vm-logic/src/logic.rs (hunk 7 above)
Comment (jakmeier) [reply runtime/near-vm-logic/src/logic.rs]: `nodes_delta` was moved before the event is emitted. I changed it to a borrow now, which makes it possible to simplify as you suggested.

File: core/o11y/src/io_tracer.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+use tracing::{span, Subscriber};
+use tracing_appender::non_blocking::{NonBlocking, WorkerGuard};
+use tracing_subscriber::fmt::MakeWriter;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::Layer;
+
+/// Tracing layer that produces a record of IO operations.
+pub struct IoTraceLayer {
+    make_writer: NonBlocking,
+    _guard: WorkerGuard,
```
Comment (nagisa) [inline core/o11y/src/io_tracer.rs:31]: I believe this misses the point of the guard somewhat. The guard here exists so that the writer can tell when the application is about to exit and to give it an opportunity to flush any internal buffers and whatnot.

If this guard ends up being stored inside the subscriber, this guard will never get dropped. If flushing at exit is not a concern, then you can drop this guard immediately after it is constructed.

Comment (jakmeier) [reply core/o11y/src/io_tracer.rs:31]: Oh I see, makes sense.

Comment (jakmeier) [reply core/o11y/src/io_tracer.rs:31]: Turns out, dropping the guard early silently shuts down the writer thread. And no output is produced.

https://docs.rs/tracing-appender/latest/src/tracing_appender/non_blocking.rs.html#270-290

So I guess I will keep it as it is. Assuming that on normal shutdown, the buffered output still gets flushed, I think I am okay with this behaviour.

Comment (nagisa) [reply core/o11y/src/io_tracer.rs:31]: Drop glue for global statics do not get run no matter how the process terminates.

File: core/o11y/src/io_tracer.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs]: Right, added a `BufWriter` here.

File: core/o11y/src/io_tracer.rs (hunk 9 above)
Comment (jakmeier) [reply core/o11y/src/io_tracer.rs]: I guess that means I should be keeping the guard properly then. :) Done in the latest commit. (after the merge)
